### ISSUE TITLE
object-shorthand should have an exception for named functions

### ISSUE THREAD

**Issue by RReverser** [eslint/eslint#2748]
In general, I want to enforce `object-shorthand` rule in all my code.

However, I also care about proper function names (`.name` property) and would like to have a [configurable] exception for cases when function name is explicitly set and is different from property name like in following case:

```javascript
Object.create(proto, {
	toString: {
		writable: true,
		configurable: true,
		// should not warn here, as I explicitly want name `toString`:
		value: function toString() {
			return this[PROXY_TARGET].toString();
		}
	}
});
```

**Comment by michaelficarra** [comment]
Why would it be warning about that case to begin with? This is a bug report, not a feature request. :+1:

**Comment by RReverser** [comment]
@michaelficarra Maybe, wasn't sure as I don't know the intentions behind `object-shorthand`. But in any case, yes, I don't expect warning here.

**Comment by michaelficarra** [comment]
More evidence that this must be a bug is that methods can't self-reference but NFEs can. This shouldn't cause a warning:

```js
!{ a: function a(){ a(); } };
```

**Comment by RReverser** [comment]
@michaelficarra Yeah, likely... but for such case I'd also check for self-reference - if it doesn't exist, then warning should be emitted.

**Comment by xjamundx** [comment]
Yeah I think we can say if there is a name it shouldn't warn yeah? I'm fine with that as a geneal exception.

**Comment by michaelficarra** [comment]
@RReverser Now that would be an enhancement.

**Comment by RReverser** [comment]
Ah, maybe.

**Comment by nzakas** [comment]
I think this is a bug. The intent is to show equivalence, and there is no equivalence here. I'd say just omit warning for any named function expression.

**Comment by RReverser** [comment]
Well, in fact there is definitely equivalence between
```javascript
{ a() { return 42 } }
```
and
```javascript
{ a: function a() { return 42 } }
```
So
> but for such case I'd also check for self-reference - if it doesn't exist, then warning should be emitted
(at least as a future improvement)

**Comment by nzakas** [comment]
There's no equivalence there, as

```
{ a() { return 42 } }
```

Is equal to

```
{ a: function() { return 42 } }
```

The function needs to be anonymous for the two to be equivalent.

For now, we should just turn off the warning of the function expression has a name. Going into whether or not the name is used is overkill IMO.

**Comment by RReverser** [comment]
Not really - according to the spec, the function's `.name` property should be equal to the name of object property, so in runtime same-named function is actually the one equivalent to concise method.

**Comment by nzakas** [comment]
The `name` is not a binding:

```
{ a() { return a; } } // no such identifier "a" - returns undefined
```

vs.

```
{ a: function a() { return a; } } // returns function ref
```

See: https://github.com/johnstonbl01/eslint-no-inferred-method-name

**Comment by RReverser** [comment]
Yes, I know and wrote about this above. I'm talking about runtime equivalent (the `name` property which is often important).

**Comment by nzakas** [comment]
I feel like we are talking in circles.

My point is that we should never warn when a property has a named function expression, as the shorthand version would not behave the same way. It doesn't matter if the name is the same as the property or not, this rule shouldn't warn for it.

If you want something that cares about function expression name being the same or different than the property, then that would be a different rule.
